src/Tester.ts

```diff
@@ -257,15 +257,19 @@ export class Tester {
             }
             // Stop handling if sent TD does not have "data".
             if (testMode == Utils.InteractionType.Event && !interaction.hasOwnProperty("data")) {
-                // receivedData === undefined if event was emitted without payload.
+                // receivedData === undefined if event was emitted without payload. This is correct as long as event has no "data" property.
                 if (receivedData === undefined) {
                     self.log("Received empty event with no data expected due to the missing data property for this event in the TD.")
                     let result = new Result(200, "Received empty event with no data expected due to the missing data property for this event in the TD.")
                     container.eventDataReport.received.push(new EventData(receivedTimeStamp, null, result))
                     return
                 }
-                self.log("Received event data with no data property for this event in the TD.")
-                let result = new Result(99, "Received data despite no data expected due to the missing data property for this event in the TD.")
+                // Received unexpected event data (no "data" property for this event in the TD).
+                self.log("Received unexpected (no data property for this event in the TD) event data: " + JSON.parse(JSON.stringify(receivedData, null, " ")))
+                let result = new Result(
+                    99,
+                    "Received unexpected (no data property for this event in the TD) event data: " + JSON.parse(JSON.stringify(receivedData, null, " "))
+                )
                 container.passed = false
                 container.eventDataReport.passed = false
                 container.eventDataReport.received.push(new EventData(receivedTimeStamp, JSON.parse(JSON.stringify(receivedData, null, " ")), result))
```

Comment:
@egekorkan
I changed this if block after noticing some inaccuracies. Could you please give feedback on this?

On this topic: I checked the old behavior before any of my changes for testAction(). There is a check if the returned value is JSON. If not, there is still a message added which has the returned non JSON value as received value. I removed this due to concerns this could be a problem when the returned value is then added to the JSON testReport. 

The reason for my question is this if block (`if (testMode == Utils.InteractionType.Event && !interaction.hasOwnProperty("data"))`). Theoretically the received data could be added to the container here but this would be before the check if it is a JSON object. If I put it behind the JSON check it could be that the result just says "no JSON format" but not that no data was expected. Is it possible to receive non JSON conformant data here? (undefined is rendered as null...) Alternatively I could keep the slightly inelegant JSON.parse(JSON.stringify(...)) solution.

Due to action output being optional in a TD i would also add this part to testAction() when you gave feedback...

Reply:
I noticed every information on a property is optional (due to PropertyAffordance being a subclass to DataSchema TD Spec §5.3.1.3). This would mean there also could be an empty property? Is this the case and if yes, I do not think the testbench can handle this. For an example the Thing Description Playground validates this TD: 
```
{
    "title": "TestServient",
    "description": "Test servient that can be used as a servient to be tested with the WoT Test Bench",
    "properties": {
        "display": {
            "observable": false,
            "readOnly": false,
            "writeOnly": false,
            "forms": [
                {
                    "href": "http://localhost:8083/TestServient/properties/display",
                    "contentType": "application/json",
                    "op": ["readproperty", "writeproperty"]
                }
            ]
        }
    },
    "id": "urn:uuid:560290c8-6490-4a58-99ca-eea9bc8c25d2",
    "@context": "https://www.w3.org/2019/wot/td/v1",
    "@type": "Thing",
    "security": ["nosec_sc"],
    "forms": [
        {
            "href": "http://localhost:8083/TestServient/all/properties",
            "contentType": "application/json",
            "op": ["readallproperties", "readmultipleproperties", "writeallproperties", "writemultipleproperties"]
        }
    ],
    "securityDefinitions": {
        "nosec_sc": {
            "scheme": "nosec"
        }
    }
}

```
![image](https://user-images.githubusercontent.com/65398308/91364043-2fed1f00-e7fe-11ea-8660-568fb7f4faec.png)

Reply:
regarding the first comment, I do not understand the question but:
>Alternatively I could keep the slightly inelegant JSON.parse(JSON.stringify(...)) solution if this solution is guaranteed to work with every possible input.

There is no such guarantee. The received data as a buffer gets converted as JSON by content serdes so that the script can use it as a JSON. We can assume that we always get JSON but this should be clearly stated.